[SeparateConstOffsetFromGEP] Don't set unsound inbounds flag

ritter-x2a opened the pull request:
The language reference says about inbounds geps that "if the
getelementptr has any non-zero indices[...] [t]he base pointer has an in
bounds address of the allocated object that it is based on [and]
[d]uring the successive addition of offsets to the address, the
resulting pointer must remain in bounds of the allocated object at each
step."

If (gep inbounds p, (a + 5)) is translated to (gep [inbounds] (gep p,
a), 5) with p pointing to the beginning of an object and a=-4, as the
example in the comments suggests, that's the case for neither of the
resulting geps. Therefore, we need to clear the inbounds flag for both
geps.

We might want to use ValueTracking to check if a is known to be
non-negative to preserve the inbounds flags.

For the AMDGPU tests with scratch instructions, removing the unsound
inbounds flag means that AMDGPUDAGToDAGISel::isFlatScratchBaseLegal sees
no NUW flag at the pointer add, which prevents generation of scratch
instructions with immediate offsets.

For SWDEV-516125.

ritter-x2a committed:
[SeparateConstOffsetFromGEP] Don't set unsound inbounds flag

The language reference says about inbounds geps that "if the
getelementptr has any non-zero indices[...] [t]he base pointer has an in
bounds address of the allocated object that it is based on [and]
[d]uring the successive addition of offsets to the address, the
resulting pointer must remain in bounds of the allocated object at each
step."

If (gep inbounds p, (a + 5)) is translated to (gep [inbounds] (gep p,
a), 5) with p pointing to the beginning of an object and a=-4, as the
example in the comments suggests, that's the case for neither of the
resulting geps. Therefore, we need to clear the inbounds flag for both
geps.

We might want to use ValueTracking to check if a is known to be
non-negative to preserve the inbounds flags.

For the AMDGPU tests with scratch instructions, removing the unsound
inbounds flag means that AMDGPUDAGToDAGISel::isFlatScratchBaseLegal sees
no NUW flag at the pointer add, which prevents generation of scratch
instructions with immediate offsets. It's not clear to me what
test/CodeGen/AMDGPU/constant-address-space-32bit.ll tests and if my fix
preserves that.

For SWDEV-516125.

ritter-x2a commented:
* **#130617** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/130617?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#130616** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/130616?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/130616?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

ritter-x2a committed:
Simplify the test case so that it is easier to handle for Alive2

ritter-x2a commented:
Here is an alive2 counter example for the transformation as it was before this PR is applied, on the test in preserve-inbounds.ll: https://alive2.llvm.org/ce/z/2unNTZ

arsenm commented on llvm/test/CodeGen/AMDGPU/constant-address-space-32bit.ll:
```diff
@@ -228,6 +228,9 @@ define amdgpu_vs i32 @load_i32_hifffffff0(ptr addrspace(6) inreg %p) #4 {
 ; GCN: v_readfirstlane_b32
```
The point of the test file is testing addrspace(6), which is just a truncated constant address space pointer

ritter-x2a replied:
So the two loads don't need to be directly adjacent in the generated code, as the test originally required here? 
Looks like they are no longer adjacent with this patch because an offset is no longer folded into the second load. There are other tests in this file checking that offsets are folded(for example `load_v8float`), so these `load_sampler` tests probably don't need to check that.

ritter-x2a replied:
Adjusted the test.

arsenm commented on llvm/test/CodeGen/AMDGPU/memory_clause.ll:
```diff
@@ -208,20 +208,26 @@ define void @mubuf_clause(ptr addrspace(5) noalias nocapture readonly %arg, ptr
+; GCN-SCRATCH-NEXT:    scratch_store_dwordx4 v19, v[14:17], off
 ; GCN-SCRATCH-NEXT:    s_setpc_b64 s[30:31]
 bb:
   %tmp = tail call i32 @llvm.amdgcn.workitem.id.x()
```
For the purpose of this test, probably best to massage the input to preserve the offset folding

ritter-x2a replied:
Adjusted the test so that the generated ISA stays the same as before.

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1103,7 +1103,7 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
   // possible. GEPs with inbounds are more friendly to alias analysis.
   // TODO(gep_nowrap): Preserve nuw at least.
-  bool GEPWasInBounds = GEP->isInBounds();
+  auto NewGEPFlags = GEPNoWrapFlags::none();
```
no auto?

ritter-x2a replied:
done

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/flat-scratch.ll:
```diff
@@ -179,12 +179,15 @@ define amdgpu_kernel void @store_load_vindex_kernel() {
 ; GFX9-NEXT:    v_sub_u32_e32 v0, 0, v0
-; GFX9-NEXT:    scratch_store_dword v1, v2, off
+; GFX9-NEXT:    v_mov_b32_e32 v2, 0
+; GFX9-NEXT:    scratch_store_dword v1, v3, off
```
change to preserve offset fold?

arsenm commented on llvm/test/CodeGen/AMDGPU/flat-scratch.ll:
```diff
@@ -682,10 +682,11 @@ define amdgpu_kernel void @store_load_vindex_kernel() {
 ; GFX9-NEXT:    scratch_store_dword v1, v2, off
 ; GFX9-NEXT:    s_waitcnt vmcnt(0)
-; GFX9-NEXT:    v_sub_u32_e32 v0, 0, v0
-; GFX9-NEXT:    scratch_load_dword v0, v0, off offset:124 glc
```
Change to preserve offset fold

ritter-x2a replied:
Isn't that offset actually just wrong here (at least pre-gfx12)? The `v_sub` computes `-workitemid*4` in `v0`, which is negative (except for 0), whereas the hardware wants an unsigned vgpr_offset.

arsenm replied:
Wrong or not, this patch is not a solution if this was an incorrect addressing mode fold

ritter-x2a replied:
Adjusted this test and the global ISel version by manually applying the unsound transformation and increasing alloca sizes so that the tests are not statically guaranteed to show UB.

arsenm replied:
Should come back to check this matching though

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/flat-scratch.ll:
```diff
@@ -195,12 +198,14 @@ define amdgpu_kernel void @store_load_vindex_kernel() {
 ; GFX10-NEXT:    s_waitcnt_vscnt null, 0x0
-; GFX10-NEXT:    scratch_load_dword v0, v1, off offset:124 glc dlc
+; GFX10-NEXT:    v_add3_u32 v1, 0, v1, v2
+; GFX10-NEXT:    scratch_load_dword v0, v1, off glc dlc
```
Add some addressing mode tests that show the net result of offset folding and the preserve / drop of these flags? Or does that go with a later patch?

ritter-x2a replied:
Added test/CodeGen/AMDGPU/fold-gep-offset.ll

ritter-x2a committed:
No auto for GEPNoWrapFlags

ritter-x2a committed:
Massage memory_clause.ll so that offsets are still generated

ritter-x2a committed:
Make sure that offsets are still folded...

...where they were folded previously, by manually applying the unsound
version of the transformation and, where necessary, increasing the
allocation size so that the transformed code is not statically
guaranteed to cause UB.

ritter-x2a committed:
Add some addressing mode tests that show the result of dropping the inbounds flag on offset folding.

arsenm commented on llvm/test/CodeGen/AMDGPU/fold-gep-offset.ll:
```diff
@@ -0,0 +1,438 @@
+  ret i32 %l
+}
+
+define i32 @lds(ptr addrspace(3) %p, i32 %i) {
```
These other address spaces aren't relevant to this issue? The different address spaces should have different tests (which we also surely already have, just scattered around inconsistently)

arsenm commented on llvm/test/CodeGen/AMDGPU/fold-gep-offset.ll:
```diff
@@ -0,0 +1,438 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+
+; RUN: llc -mtriple=amdgcn -mcpu=gfx90a -mattr=-enable-flat-scratch < %s | FileCheck --check-prefixes=GFX90A,GFX90A-FSOFF %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx90a -mattr=+enable-flat-scratch < %s | FileCheck --check-prefixes=GFX90A,GFX90A-FSON %s
```
Call the off case MUBUF vs. FLATSCR

arsenm commented on llvm/test/CodeGen/AMDGPU/fold-gep-offset.ll:
```diff
@@ -0,0 +1,438 @@
+; This test documents when memory addresses with constant offset components can
+; be folded into memory accesses with immediate offsets.
+
+define i32 @flat(ptr addrspace(0) %p, i32 %i) {
```
Test name is generic, and it's only the positive fold case. The goal is to stress the negative case where it's illegal to fold the offset based on flags

ritter-x2a replied:
As it is, this PR only affects flat scratch (and apparently SMEM, as the const32 test suggests, not sure why it makes a difference there). Ideally, it should also affect generic flat instructions, since we currently miscompile those when the final GEP is not inbounds, but changing that requires ISel changes for which I haven't opened a PR yet.

In the new version, the add no longer is `nuw`, so that the flat case is indeed a miscompile. I also only kept the flat and scratch cases and renamed them.
Should this test also check the cases where `%p` and/or `%i` are uniform?

ritter-x2a committed:
fold-gep-offset.ll: Drop unrelated addrspaces, drop nuw, more expressive names, use MUBUF&FLATSCR.

ritter-x2a committed:
Adjust fold-gep-offset.ll to trunk changes

arsenm commented on llvm/test/CodeGen/AMDGPU/fold-gep-offset.ll:
```diff
@@ -0,0 +1,159 @@
+
+; FIXME the offset here should not be folded: if %p points to the beginning of
+; scratch or LDS and %i is -1, a folded offset crashes the program.
+define i32 @flat_offset_maybe_oob(ptr addrspace(0) %p, i32 %i) {
```
Maybe drop the explicit addrspace(0)s

ritter-x2a replied:
Done.

ritter-x2a committed:
Drop explicit addrspace(0)

ritter-x2a commented:
### Merge activity

* **Mar 18, 7:28 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/130616).
